tests/test/builtin_impls.rs

```diff
@@ -0,0 +1,604 @@
+        goal { i16: Sized } yields { "Unique" }
+        goal { i32: Sized } yields { "Unique" }
+        goal { i64: Sized } yields { "Unique" }
+        goal { i128: Sized } yields { "Unique" }
+        goal { isize: Sized } yields { "Unique" }
+        goal { u8: Sized } yields { "Unique" }
+        goal { u16: Sized } yields { "Unique" }
+        goal { u32: Sized } yields { "Unique" }
+        goal { u64: Sized } yields { "Unique" }
+        goal { u128: Sized } yields { "Unique" }
+        goal { usize: Sized } yields { "Unique" }
+        goal { f32: Sized } yields { "Unique" }
+        goal { f64: Sized } yields { "Unique" }
+        goal { bool: Sized } yields { "Unique" }
+        goal { char: Sized } yields { "Unique" }
+    }
+}
+
+#[test]
+fn dyn_to_dyn_unsizing() {
+    test! {
+        program {
+            #[lang(unsize)]
+            trait Unsize<T> {}
+
+            trait Principal {}
+            trait OtherPrincipal {}
+            trait GenericPrincipal<T> {
+                type Item;
+            }
+
+            #[auto]
+            trait Auto1 {}
+
+            #[auto]
+            trait Auto2 {}
+
+            #[auto]
+            trait Auto3 {}
+        }
+
+        goal {
+            forall<'a> {
+                forall<'b> {
+                    dyn Principal + 'a: Unsize<dyn Principal + 'b>
+                }
+            }
+        } yields {
+            "Unique; substitution [], lifetime constraints [InEnvironment { environment: Env([]), goal: '!1_0 == '!2_0 }]"
+        }
+
+        goal {
+            forall<'a> {
+                forall<'b> {
+                    dyn Principal + Auto1 + Auto2 + 'a: Unsize<dyn Principal + Auto1 + 'b>
+                }
+            }
+        } yields {
+            "Unique; substitution [], lifetime constraints [InEnvironment { environment: Env([]), goal: '!1_0 == '!2_0 }]"
+        }
+
+        goal {
+            forall<'a> {
+                forall<'b> {
+                    dyn Principal + Auto1 + Auto2 + 'a: Unsize<dyn Principal + Auto1 + Auto3 + 'b>
+                }
+            }
+        } yields {
+            "No possible solution"
+        }
+
+        goal {
+            forall<'a> {
+                forall<'b> {
+                    dyn Principal + 'a: Unsize<dyn OtherPrincipal + 'b>
+                }
+            }
+        } yields {
+            "No possible solution"
+        }
+
+        goal {
+            forall<'a> {
+                forall<'b> {
+                    dyn GenericPrincipal<u64, Item = u64> + 'a: Unsize<dyn GenericPrincipal<u64, Item = u64> + 'b>
+                }
+            }
+        } yields {
+            "Unique"
+        }
+
+        goal {
+            forall<'a> {
+                forall<'b> {
+                    dyn GenericPrincipal<u32, Item = u32> + 'a: Unsize<dyn GenericPrincipal<u32, Item = u64> + 'b>
+                }
+            }
+        } yields {
+            "No possible solution"
+        }
+    }
+}
```

Comment:
Can you add

checking that using the same lifetime is fine
```
goal {
  forall<'a> {
    dyn Principal + 'a: Unsize<dyn Principal + 'a>
  }
} yields {
  "Unique; substitution [], lifetime constraints []"
}
```
check when source has subset of auto traits
```
goal {
  forall<'a> {
    forall<'b> {
      dyn Principal + Auto2 + 'a: Unsize<dyn Principal + Auto1 +Auto2 + 'b>
    }
  }
} yields {
  "No possible Solution"
}
```
checking auto trait order
```
goal {
  forall<'a> {
    forall<'b> {
      dyn Principal + Auto1 + Auto2 + 'a: Unsize<dyn Principal + Auto2 + Auto1 + 'b>
    }
  }
} yields {
  "Unique; substitution [], lifetime constraints [InEnvironment { environment: Env([]), goal: '!1_0 == '!2_0 }]"
}
```
checking when the principal trait is missing
```
goal {
  forall<'a> {
    forall<'b> {
      dyn Principal + Auto1 + 'a: Unsize<dyn Auto1 + 'b>
    }
  }
} yields {
  "No possible solution"
}
```
checking if the principal isn't the first trait
```
goal {
  forall<'a> {
    forall<'b> {
      dyn Principal + Auto1 + 'a: Unsize<dyn Auto1 + Principal + 'b>
    }
  }
} yields {
  "Unique; substitution [], lifetime constraints [InEnvironment { environment: Env([]), goal: '!1_0 == '!2_0 }]"
}
```